### PR TITLE
scripts: dts: disable checks for cyclic dependencies in the devicetree

### PR DESCRIPTION
**Pull request by benediktibk**
Disable the checks for cyclic dependencies in the devicetree, as it is not really necessary to enforce this.

Cycles in the device tree can occur for instance with SPI port expanders, as they by themselves can be used for the chip select of other devices on the same SPI bus.

	&spi1 {
		cs-gpios = <&gpiof 10 GPIO_ACTIVE_LOW>,			
			   <&temp_port_expander 0 GPIO_ACTIVE_LOW>;
		...

		temp_port_expander: temp_port_expander@0 {
			compatible = "microchip,mcp23sxx";
			...
		};

		temp_adc1: temp_adc1@1 {
			compatible = "ti,ads114s08";
			...
		};
	};

With these changes so far only applications are affected, which have a cycle in their devicetree dependency graph.

As far as I can see it the scc_order is only used to determine the device ordinals. I suspect these were originally implicitly used for the device initialization order as well. By now this is defined via initialization levels and priorities, hence it is not necessary anymore to enforce cycle free device tree dependency graphs.

I suspect the device ordinals could also be random (or be given by their first appearance in the devicetree) and the scc_order could be removed completely. But this would of course affect all existing applications, therefore I refrained from changing this.

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
Figure out how to this only when a board specifically needs this since its extremely rare and want to maintain the check.

**Comment by benediktibk** [comment]
> I would suggest looking at how to do this specifically for your board/platform. In general we want to maintain this check. Its extremely rare to have a case that is cyclic.
> 
> So I thinking adding some flag that can be passed to the script to skip the check would probably be best.

I'm fine with disabling the check only with a flag.

But could you please explain why you would like to keep the check anyway?

**Comment by benediktibk** [comment]
Made it work like this:

     west build --board native_posix -d build zephyr/samples/philosophers -- -DEXTRA_GEN_DEFINES_ARGS=--disable-cyclic-dependency-check

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
The commit logs are explaining what the patches do, but not why they are being added.

Before jumping into solutions mode for such a major change, I think this needs a better justification and a clearer story on how it would impact things like https://github.com/zephyrproject-rtos/zephyr/issues/22545 that are widely discussed and agreed as desirable.

I know this is not changing the default behavior, but as soon as this gets introduced, people will start using it upstream, and that will cause problems when we finally get to the use cases we want to enable where the devicetree dependencies are used to order boot-time device initialization etc.

We also want to be able to track dependencies for things like operating-system-assisted power management where we don't want to put a device to sleep if something that depends on it is busy, etc.

TL;DR we need a clear model of how dependencies will be managed in a world where this is possible

**Comment by benediktibk** [comment]
I get your concerns @mbolivar-nordic , but I think the use case is pretty clearly described in the PR description. Hence, there will be hardware, and in my case there is already one, which will have cyclic dependencies in the device tree. If you are planning on using the device ordinals for the device initialization order I understand that you would like to avoid such a flag which disables this check. Therefore I suppose my approach with disabling this check based upon a flag won't be accepted at all?

The basic question is then which of the devices in a cycle should be initialized first? In my case, with the port expander on a bus being used for the chip selects of other devices on the same bus, the solution is clear: Bus controller -> port expander on bus -> other devices on the bus. But of course it will be harder to detect this automatically, as in the graph this is represented as two nodes depending on each other (bus controller on bus device).

Would it make sense to implement a general precedence of bus controllers over bus devices? This would of course imply passing through this information to the graph from the devicetree. And I suppose it could become more complex for instance in:

    bus_controller_A {
        phandle = <&bus_controller_B>;
        bus_device_A {
        };
    };

    bus_controller_B {
        phandle = <&bus_device_A>;
    };

Maybe it would help to phrase it differently, saying a child (or grand child) must always be initialized after its parents or grandparents, as fallback for cycles?

**Comment by mbolivar-nordic** [comment]
> I think the use case is pretty clearly described in the PR description

The PR description will get lost. Commit logs are the permanent record. Please read the contribution guidelines; this is not optional.

> The basic question is then which of the devices in a cycle should be initialized first?

I would say the basic question is how we want to refine the "depends on" relationship in the presence of cycles in a way that addresses use cases like this without precluding the future work that people want to do (not just boot initialization, but power mgmt too)

**Comment by benediktibk** [comment]
> > I think the use case is pretty clearly described in the PR description
> 
> The PR description will get lost. Commit logs are the permanent record. Please read the contribution guidelines; this is not optional.
> 
> > The basic question is then which of the devices in a cycle should be initialized first?
> 
> I would say the basic question is how we want to refine the "depends on" relationship in the presence of cycles in a way that addresses use cases like this without precluding the future work that people want to do (not just boot initialization, but power mgmt too)

Okay, I will include the reasoning behind the change in the commit message. But first we will have to figure what we want to aim for.

Do you think the backup order of level depth (parent > child) is reasonable in the context of device initialization and power management?

**Comment by benediktibk** [comment]
Won't continue down this road, reason in https://github.com/zephyrproject-rtos/zephyr/issues/56920#issuecomment-1514621591.
